r/R/dplyr-funcs-type.R

```diff
@@ -76,6 +76,64 @@ register_bindings_type_cast <- function() {
   register_binding("as.numeric", function(x) {
     Expression$create("cast", x, options = cast_options(to_type = float64()))
   })
+  register_binding("as.Date", function(x,
+                                       format = NULL,
+                                       tryFormats = "%Y-%m-%d",
+                                       origin = "1970-01-01",
+                                       tz = "UTC") {
+
+    if (call_binding("is.Date", x)) {
+      # base::as.Date() first converts to the desired timezone and then extracts
+      # the date, which is why we need to go through timestamp() first
+      return(x)
+
+    # cast from POSIXct
+    } else if (call_binding("is.POSIXct", x)) {
+      if (tz == "UTC") {
+        interim_x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+      } else {
+        abort("`as.Date()` with a timezone different to 'UTC' is not supported in Arrow")
+      }
+
+    # cast from character
+    } else if (call_binding("is.character", x)) {
+      # this could be improved with tryFormats once strptime returns NA and we
+      # can use coalesce - https://issues.apache.org/jira/browse/ARROW-15659
+      # TODO revisit once https://issues.apache.org/jira/browse/ARROW-15659 is done
+      if (is.null(format)) {
+        if (length(tryFormats) == 1) {
+          format <- tryFormats[1]
+        } else {
+          abort("`as.Date()` with multiple `tryFormats` is not supported in Arrow yet")
+        }
+      }
+      # if x is not an expression (e.g. passed as filter), convert it to one
+      if (!inherits(x, "Expression")) {
+        x <- build_expr("cast", x, options = cast_options(to_type = type(x)))
+      }
```

Comment:
Could you describe a little bit more about what cases this is catching? 

I ask because I suspect that this is catching with the message we are testing:

```
expect_error(
    test_df %>%
      arrow_table() %>%
      mutate(date_char = date(character_ymd_var)) %>%
      collect(),
    regexp = "Unsupported cast from string to date32 using function cast_date32"
  )
```

I was expecting something like this: 

```
> tab <- arrow_table(data.frame(time = "2020-01-01 00:01:00", date = "2020-01-01"))
> tab %>% mutate(strptime(time, format = "%Y-%m-%d")) %>% collect()
Error in `handle_csv_read_error()`:
! Invalid: Failed to parse string: '2020-01-01 00:01:00' as a scalar of type timestamp[ms]
Run `rlang::last_error()` to see where the error occurred.
```

So I suspect that this `if()` is catching this before that, which I'm a little surprised by. So anyway, what cases are we expecting here?

Reply:
And separately — we should make a Jira if one doesn't already exist to see if we can allow `strptime(time, format = "%Y-%m-%d")` to work (which does in R)

```
> strptime("2020-01-01 00:01:00", "%Y-%m-%d")
[1] "2020-01-01 CST"
```

Reply:
I introduced that step (converting a character `x` to `Expression`) as a way of dealing with the errors we were seeing with `filter()`. In `filter.arrow_dplyr_query()` we get the following error after tidy-eval-ing the filter expression:
```r
> filters <- lapply(filts, arrow_eval, arrow_mask(.data))
> filters
[[1]]
[1] "Expression arguments must be Expression objects"
attr(,"class")
[1] "try-error"
attr(,"condition")
<assertError: Expression arguments must be Expression objects>
``` 

where `filts` is:
```r
> filts
<list_of<quosure>>

[[1]]
<quosure>
expr: ^ts >= as.Date("2015-05-04")
env:  0x297eaa768
``` 

and the original test is:
```r
# ds is an open dataset
ds %>%
      filter(ts >= as.Date("2015-05-04")) %>%
      filter(part == 1) %>%
      select(ts) %>%
      collect()
```

As far as I understand it, it fails due to `"2015-05-04"` being a character and not an `Expression` object. Inside the `filter. ...` method:
```r
> b <- arrow_eval(filts[[1]], arrow_mask(.data))
> b
[1] "Expression arguments must be Expression objects"
attr(,"class")
[1] "try-error"
attr(,"condition")
<assertError: Expression arguments must be Expression objects>
```
Hence my approach: if we somehow get to that point in the function body and, for some reason, character `x` is not yet an `Expression, make it into one.

Reply:
The additional casting step does not impact `date()` (the unit test you copied), only `as.Date()`.

Reply:
> And separately — we should make a Jira if one doesn't already exist to see if we can allow `strptime(time, format = "%Y-%m-%d")` to work (which does in R)
> 
> ```
> > strptime("2020-01-01 00:01:00", "%Y-%m-%d")
> [1] "2020-01-01 CST"
> ```

Done. [ARROW-15813](https://issues.apache.org/jira/browse/ARROW-15813)